discovery.md

```diff
@@ -537,17 +563,118 @@ When encoded in JSON, the response format MUST adhere to the following:
+  location of the Service was that action upon. The response body
+  MUST include the Service that was being operated on, unless it
+  has been deleted, then the response body MUST be empty.
+- `202 Accepted` indicates that the original request is still being processed.
+  The response body MAY be empty.
+- `406 Not Acceptable` to indicate that the original request failed to be
+  processed correctly. The HTTP response Body SHOULD include additional
+  information as to why it failed.
+
+Other responses are allowed, but not defined by this specification, however
+they are related to the `GET` itself and not the original request.
+
+How long a Discovery Endpoint supports requests to the status endpoint
+is an implementation choice, however, it MUST be available immediately
+upon return of the `202 Accepted` from the original request.
+
+#### `POST /services`
+
+This MUST add a Service to the list of Services available at this Discovery
+Endpoint. The Body of the request message MUST contain a valid single Service
+Entry, with the exception that it MUST NOT include an `id` attribute.
+The `id` MUST be generated by the Discovery Endpoint.
+
+The follow responses are defined by this specification:
+- `201 Created` if the new Service was added. The response MUST include
+  an HTTP `Location` Header with a value that points to the new Service.
+  The HTTP Response Body MUST include the JSON represenation of the
+  newly created Service.
+- `404 Not Found` if there is no Service with the specified `id`.
+
+Other responses are allowed, but not defined by this specification.
+
+#### `PUT /services/{id}`
+
+This MUST add or update a Service in a Discovery Endpoint. If a Service
+already exists with the specified `id` then this MUST attempt to update that
+Service. If it does not exist, then a new Service MUST be added with that `id`.
+
+The Body of the request message MUST contain a single Service
+definition. Unlike the `POST` operation, the Service MUST contain an `id`
+attribute that matches the `{id}` in the `PUT` URL.
+
+In the case of a new Service being created, the sender MAY choose to include
+the `epoch` attribute. It will do so only in cases where it is doing an
+"import" type of operation where preserving the previously defined value
+is needed. For "create" type of operations the sender MUST NOT include it.
+
+When a Service with the specified `id` already exists, the client
+MAY choose to include the `epoch` attribute. If present then the Discovery
+Endpoint MUST ensure that the value on the incoming request matches the
+current value on the Service. If they do not match then an error MUST be
+generated and the Service MUST NOT be updated. However, if the incoming
+request does not include this attribute then this verification MUST NOT
+be done.
+
+Additionally, when a Service with the specified `id` aleady exists, the
+`epoch` value MUST be updated with a larger value, indicating that the
+Service has been updated.
```

Comment:
I'm a bit confused by this and the preceding paragraph. It seems to be saying that when updating an existing service, the singular `epoch` attribute must be both equal to AND larger than the current value.

Perhaps the intent is that the epoch value is incremented by the Discovery Endpoint, not the requester, and the new value is returned in the response? In this case, is there no way for the requester to choose an epoch value?

Reply:
yes and yes. I'll make it clearer that the D.E. is the one doing the update. The only way a client can set an epoch value is during an import.  But how a D.E. knows it's an import vs an update is an open question as of now.

Reply:
Based on this and #699, it feels like this "update" operation would be more clear semantically if it is actually a creation of an epoch resource owned by the service resource. Update is then a POST to the service's resource path that returns as its Location the resource path of the epoch that was created:

`POST /services/{id}`

This also solves the problem of import vs update, because you can allow PUT to a specific epoch:

`PUT /services/{id}/{epoch}`

Reply:
I realized there's another issue just for this #701 so I cross-posted the above there.

Reply:
I'd like to deal with your suggestion in a separate issue/PR if that's ok. Mainly because I think the proposal will require a lot of discussion and I'd like to get _something_ in that we can morph and test-out rather than not have anything at all for potentially several more weeks.